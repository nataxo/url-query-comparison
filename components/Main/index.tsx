import {useState, useEffect} from 'react';
import {useRouter} from 'next/router';
import {compareUrls, parseUrl} from '../../helpers/url';
import {Diff, Eq} from '../../types';
import View from './view';

const compare = (first: string, second: string, ignore?: string) => {
    const ignoreParams = ignore?.trim().replace(' ', '').split(',') ?? [];

    try {
        localStorage.setItem('ignoreParams', ignore ?? '');
    } catch (e) { }

    return compareUrls(
        first.trim(),
        second.trim(),
        ignoreParams,
    );
};

const Main = () => {
    const [firstUrl, setFirstUrl] = useState('');
    const [secondUrl, setSecondUrl] = useState('');
    const [ignoreParams, setIgnoreParams] = useState('');
    const [difference, setDifference] = useState<Diff[]>([]);
    const [equal, setEqual] = useState<Eq[]>([]);
    const router = useRouter();

    const changeAppState = ({first = '', second = '', ignore = ''}) => {
        const ignoreParams =
            ignore ?? localStorage.getItem('ignoreParams') ?? '';
        const { diff, eq } = compare(first, second, ignoreParams);

        setFirstUrl(first);
        setSecondUrl(second);
        setIgnoreParams(ignoreParams);
        setDifference(diff);
        setEqual(eq);
    };

    useEffect(() => {
        // init app state
        changeAppState(router.query);

        const handleRouteChange = (url: string) => {
            const {searchParams} = parseUrl(url);
            changeAppState({
                first: searchParams.get('first') || '',
                second: searchParams.get('second') || '',
                ignore: searchParams.get('ignore') || '',
            });
        };

        router.events && router.events.on('routeChangeComplete', handleRouteChange);
        return () => {
            router.events && router.events.off('routeChangeComplete', handleRouteChange);
        };
    }, []);

    const changeQueryParams = (query = {}) => {
        router.push({pathname: '/', query});
    };

    const onReset = () => {
        changeQueryParams();
    };

    const onSubmit = (e: React.SyntheticEvent) => {
        e.preventDefault();

        changeQueryParams({
            first: firstUrl,
            second: secondUrl,
            ignore: ignoreParams,
        });
    };

    return (
        <View
            firstUrl={firstUrl}
            secondUrl={secondUrl}
            ignoreParams={ignoreParams}
            equal={equal}
            difference={difference}
            onFirstUrlChange={setFirstUrl}
            onSecondUrlChange={setSecondUrl}
            onIgnoreParamsChange={setIgnoreParams}
            onSubmit={onSubmit}
            onReset={onReset}
        />
    );
};

export default Main;
