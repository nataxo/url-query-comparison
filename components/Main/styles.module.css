.main {
    display: flex;
    flex-direction: column;

    padding: 24px;
    max-width: 100%;
}

.h1,
.h2 {
    font-weight: bold;
    font-size: 18px;
    margin-bottom: 12px;
}

.h1 {
    margin: 0 0 24px;
}

.info {
    color: #666;
}

@media screen and (max-width: 599px) {
    .main {
        padding: 20px 10px;
        width: 100%;
        overflow-x: hidden;
    }

    .h1 {
        margin: 0 0 12px;
    }
}
