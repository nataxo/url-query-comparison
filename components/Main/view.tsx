import Snippet from '../Snippet';
import Form, {Props as FormPropsType} from '../Form';
import Table from '../Table';
import Footer from '../Footer';

import {Diff, Eq} from '../../types';

import styles from './styles.module.css';

type Props = {
    equal: Eq[],
    difference: Diff[],
} & FormPropsType;
const View = ({equal, difference, ...formProps}: Props) => {
    return (
        <div>
            <main className={styles.main}>
                <h1 className={styles.h1}>Compare Urls</h1>
                <Snippet>
                    <Form {...formProps} />
                </Snippet>
                <Snippet>
                    <h2 className={styles.h2}>Difference</h2>
                    {difference.length > 0 ? (
                        <Table
                            titles={['Param', 'First Url', 'Second Url']}
                            values={difference}
                        />
                    ) : (
                        <div className={styles.info}>No differences</div>
                    )}
                </Snippet>
                {equal.length > 0 && (
                    <Snippet>
                        <h2 className={styles.h2}>Equal</h2>
                        <Table titles={['Param', 'Value']} values={equal} />
                    </Snippet>
                )}
            </main>
            <Footer />
        </div>
    );
};

export default View;
