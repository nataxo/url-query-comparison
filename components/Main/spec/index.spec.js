import Main from '../';
import {shallow, mount} from 'enzyme';
import { RouterContext } from 'next/dist/next-server/lib/router-context';

describe('Main content', () => {
    it('should match snapshot by default', () => {
        const component = shallow(<Main />);
        expect(component).toMatchSnapshot();
    });
    it('should match snapshot with query params', async done => {
        const query = {
            first: 'https://github.com?a=1',
            second: 'https://github.com?b=1',
        };

        const component = mount(
            <RouterContext.Provider value={{query}}>
                <Main />
            </RouterContext.Provider>,
        );
        setImmediate(() => {
            component.update();
            done(expect(component).toMatchSnapshot());
        });
    });
});
