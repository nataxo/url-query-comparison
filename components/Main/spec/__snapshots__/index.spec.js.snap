// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main content should match snapshot by default 1`] = `
<View
  difference={Array []}
  equal={Array []}
  firstUrl=""
  ignoreParams=""
  onFirstUrlChange={[Function]}
  onIgnoreParamsChange={[Function]}
  onReset={[Function]}
  onSecondUrlChange={[Function]}
  onSubmit={[Function]}
  secondUrl=""
/>
`;

exports[`Main content should match snapshot with query params 1`] = `
<Main>
  <View
    difference={
      Array [
        Array [
          "a",
          "1",
          null,
        ],
        Array [
          "b",
          null,
          "1",
        ],
      ]
    }
    equal={Array []}
    firstUrl="https://github.com?a=1"
    ignoreParams=""
    onFirstUrlChange={[Function]}
    onIgnoreParamsChange={[Function]}
    onReset={[Function]}
    onSecondUrlChange={[Function]}
    onSubmit={[Function]}
    secondUrl="https://github.com?b=1"
  >
    <div>
      <main
        className="___main_16vyg_1"
      >
        <h1
          className="___h1_16vyg_1"
        >
          Compare Urls
        </h1>
        <Snippet>
          <div
            className="___div_1tt9o_1"
          >
            <_default
              firstUrl="https://github.com?a=1"
              ignoreParams=""
              onFirstUrlChange={[Function]}
              onIgnoreParamsChange={[Function]}
              onReset={[Function]}
              onSecondUrlChange={[Function]}
              onSubmit={[Function]}
              secondUrl="https://github.com?b=1"
            >
              <form
                onSubmit={[Function]}
              >
                <_default
                  id="first-url"
                  label="First Url"
                  onChange={[Function]}
                  placeholder="first.com?alpha=1&beta=2"
                  value="https://github.com?a=1"
                >
                  <div
                    className="___Wrapper_1h3a9_1"
                  >
                    <label
                      className="___label_1h3a9_1"
                      htmlFor="first-url"
                    >
                      First Url
                    </label>
                    <textarea
                      className="___textarea_1h3a9_1"
                      id="first-url"
                      onChange={[Function]}
                      placeholder="first.com?alpha=1&beta=2"
                      rows={2}
                      value="https://github.com?a=1"
                    />
                  </div>
                </_default>
                <_default
                  id="second-url"
                  label="Second Url"
                  onChange={[Function]}
                  placeholder="second.com?beta=3&gamma=2"
                  value="https://github.com?b=1"
                >
                  <div
                    className="___Wrapper_1h3a9_1"
                  >
                    <label
                      className="___label_1h3a9_1"
                      htmlFor="second-url"
                    >
                      Second Url
                    </label>
                    <textarea
                      className="___textarea_1h3a9_1"
                      id="second-url"
                      onChange={[Function]}
                      placeholder="second.com?beta=3&gamma=2"
                      rows={2}
                      value="https://github.com?b=1"
                    />
                  </div>
                </_default>
                <_default
                  id="ignore-params"
                  label="Ignore params"
                  onChange={[Function]}
                  placeholder="beta,zetta"
                  rows={1}
                  value=""
                >
                  <div
                    className="___Wrapper_1h3a9_1"
                  >
                    <label
                      className="___label_1h3a9_1"
                      htmlFor="ignore-params"
                    >
                      Ignore params
                    </label>
                    <textarea
                      className="___textarea_1h3a9_1"
                      id="ignore-params"
                      onChange={[Function]}
                      placeholder="beta,zetta"
                      rows={1}
                      value=""
                    />
                  </div>
                </_default>
                <div
                  className="___ButtonGroup_uilxe_1"
                >
                  <_default
                    type="submit"
                  >
                    <button
                      className="___button_3fpq1_1 __type_3fpq1_1 __type_submit_3fpq1_1"
                      type="submit"
                    >
                      Compare
                    </button>
                  </_default>
                  <_default
                    onClick={[Function]}
                    type="reset"
                  >
                    <button
                      className="___button_3fpq1_1 __type_3fpq1_1 __type_reset_3fpq1_1"
                      onClick={[Function]}
                      type="reset"
                    >
                      Clean
                    </button>
                  </_default>
                </div>
              </form>
            </_default>
          </div>
        </Snippet>
        <Snippet>
          <div
            className="___div_1tt9o_1"
          >
            <h2
              className="___h2_16vyg_1"
            >
              Difference
            </h2>
            <_default
              titles={
                Array [
                  "Param",
                  "First Url",
                  "Second Url",
                ]
              }
              values={
                Array [
                  Array [
                    "a",
                    "1",
                    null,
                  ],
                  Array [
                    "b",
                    null,
                    "1",
                  ],
                ]
              }
            >
              <table
                className="___table_1qwty_1"
              >
                <thead
                  className="___thead_1qwty_1"
                >
                  <tr
                    className="___tr_1qwty_1"
                  >
                    <th
                      className="___th_1qwty_1"
                      key="0"
                    >
                      Param
                    </th>
                    <th
                      className="___th_1qwty_1"
                      key="1"
                    >
                      First Url
                    </th>
                    <th
                      className="___th_1qwty_1"
                      key="2"
                    >
                      Second Url
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    className="___tr_1qwty_1"
                    key="0"
                  >
                    <td
                      className="___td_1qwty_1"
                      key="0"
                    >
                      <CopiableText
                        value="a"
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            a
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                    <td
                      className="___td_1qwty_1"
                      key="1"
                    >
                      <CopiableText
                        value="1"
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            1
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                    <td
                      className="___td_1qwty_1"
                      key="2"
                    >
                      <CopiableText
                        value={null}
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            -
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                  </tr>
                  <tr
                    className="___tr_1qwty_1"
                    key="1"
                  >
                    <td
                      className="___td_1qwty_1"
                      key="0"
                    >
                      <CopiableText
                        value="b"
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            b
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                    <td
                      className="___td_1qwty_1"
                      key="1"
                    >
                      <CopiableText
                        value={null}
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            -
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                    <td
                      className="___td_1qwty_1"
                      key="2"
                    >
                      <CopiableText
                        value="1"
                      >
                        <button
                          className="___button_6xnkp_1"
                          onClick={[Function]}
                        >
                          <span
                            className="___Text_6xnkp_1"
                          >
                            1
                          </span>
                        </button>
                      </CopiableText>
                    </td>
                  </tr>
                </tbody>
              </table>
            </_default>
          </div>
        </Snippet>
      </main>
      <Footer>
        <footer
          className="___footer_1az0p_1"
        >
          <a
            className="___a_1az0p_1"
            href="https://github.com/nataxo/url-query-comparison"
            rel="nofollow noreferrer noopen"
            target="_blank"
            title="Github: nataxo/url-query-comparison"
          >
            <Octocat>
              <svg
                height="20"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 0C5.36987 0 0 5.36987 0 12C0 17.6395 3.86962 22.32 9.09038 23.6395C9.03013 23.4894 9.00051 23.279 9.00051 23.0696V21.0291H7.50026C6.69038 21.0291 5.93974 20.6696 5.60987 20.0089C5.21974 19.2889 5.15949 18.1787 4.16987 17.4883C3.86962 17.2483 4.10962 17.0083 4.43949 17.038C5.06962 17.2177 5.57923 17.6385 6.05923 18.2676C6.53923 18.8977 6.74962 19.0478 7.64936 19.0478C8.06911 19.0478 8.72885 19.0182 9.32936 18.9273C9.65923 18.0868 10.2291 17.3372 10.9195 16.9777C6.89974 16.4977 4.97974 14.5174 4.97974 11.8172C4.97974 10.6468 5.48936 9.53668 6.32987 8.57668C6.06026 7.6463 5.69974 5.7263 6.45038 4.97668C8.25089 4.97668 9.33038 6.14706 9.6 6.4463C10.4997 6.14604 11.4904 5.9663 12.5096 5.9663C13.5595 5.9663 14.5195 6.14604 15.4192 6.4463C15.6889 6.14604 16.7694 4.97668 18.5689 4.97668C19.2889 5.69668 18.959 7.6463 18.6587 8.57668C19.4992 9.50706 19.9792 10.6468 19.9792 11.8172C19.9792 14.5174 18.0889 16.4977 14.0997 16.9174C15.2099 17.4873 15.9901 19.1071 15.9901 20.3071V23.04C15.9901 23.1299 15.9605 23.2197 15.9605 23.3096C20.64 21.6899 24 17.2504 24 12C24 5.36987 18.6301 0 12 0Z"
                  fill="#000"
                />
              </svg>
            </Octocat>
          </a>
        </footer>
      </Footer>
    </div>
  </View>
</Main>
`;
